chain/chain/src/update_shard.rs

```diff
@@ -0,0 +1,335 @@
+use crate::crypto_hash_timer::CryptoHashTimer;
+use crate::types::{
+    ApplySplitStateResult, ApplySplitStateResultOrStateChanges, ApplyTransactionResult,
+    RuntimeAdapter, RuntimeStorageConfig,
+};
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::challenge::ChallengesResult;
+use near_primitives::hash::CryptoHash;
+use near_primitives::receipt::Receipt;
+use near_primitives::sandbox::state_patch::SandboxStatePatch;
+use near_primitives::shard_layout::ShardUId;
+use near_primitives::sharding::ShardChunk;
+use near_primitives::types::chunk_extra::ChunkExtra;
+use near_primitives::types::{Balance, BlockHeight, Gas, StateChangesForSplitStates, StateRoot};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Information about block in for which shard is updated.
+/// Use cases include:
+/// - queries to epoch manager
+/// - allowing contracts to get current chain data
+#[derive(Clone, Debug)]
+pub struct BlockContext {
+    pub block_hash: CryptoHash,
+    pub prev_block_hash: CryptoHash,
+    pub challenges_result: ChallengesResult,
+    pub block_timestamp: u64,
+    pub gas_price: Balance,
+    pub height: BlockHeight,
+    pub random_seed: CryptoHash,
+    pub is_first_block_with_chunk_of_version: bool,
+}
+
+#[derive(Debug)]
+pub struct SameHeightResult {
+    pub(crate) shard_uid: ShardUId,
+    pub(crate) gas_limit: Gas,
+    pub(crate) apply_result: ApplyTransactionResult,
+    pub(crate) apply_split_result_or_state_changes: Option<ApplySplitStateResultOrStateChanges>,
+}
+
+#[derive(Debug)]
+pub struct DifferentHeightResult {
+    pub(crate) shard_uid: ShardUId,
+    pub(crate) apply_result: ApplyTransactionResult,
+    pub(crate) apply_split_result_or_state_changes: Option<ApplySplitStateResultOrStateChanges>,
+}
+
+#[derive(Debug)]
+pub struct SplitStateResult {
+    // parent shard of the split states
+    pub(crate) shard_uid: ShardUId,
+    pub(crate) results: Vec<ApplySplitStateResult>,
+}
+
+#[derive(Debug)]
+pub enum ApplyChunkResult {
+    SameHeight(SameHeightResult),
+    DifferentHeight(DifferentHeightResult),
+    SplitState(SplitStateResult),
+}
+
+/// Reason to update a shard when new block appears on chain.
+pub enum ShardUpdateReason {
+    /// Block has a new chunk for the shard.
+    /// Contains chunk itself and all new incoming receipts to the shard.
+    NewChunk(ShardChunk, Vec<Receipt>),
+    /// Block doesn't have a new chunk for the shard.
+    /// Instead, previous chunk header is copied.
+    /// Contains result of shard update for previous block.
+    OldChunk(ChunkExtra),
+    /// See comment to `split_state_roots` in `Chain::get_apply_chunk_job`.
+    /// Process only state changes caused by resharding.
+    StateSplit(StateChangesForSplitStates),
+}
+
+/// Information about shard to update.
+pub struct ShardInfo {
```

Comment:
nit: maybe let's call it `ShardContext` to be consistent with `BlockContext`? also `ShardInfo` already exists in our codebase, which is not a big problem, but it would be nice to avoid having duplicate names I guess